kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -486,57 +545,195 @@ func (c *ipamController) onBlockDeleted(key model.BlockKey) {
+	// Keep track of various counts so that we can report them as metrics. These counts track legacy metrics by node.
+	legacyBlocksByNode := map[string]int{}
+	legacyBorrowedIPsByNode := map[string]int{}
 
 	// Iterate blocks to determine the correct metric values.
-	for _, kvp := range c.allBlocks {
-		b := kvp.Value.(*model.AllocationBlock)
-		if b.Affinity != nil {
-			n := strings.TrimPrefix(*b.Affinity, "host:")
-			blocksByNode[n]++
-		} else {
-			// Count blocks with no affinity as a pseudo-node.
-			blocksByNode["no_affinity"]++
-		}
-
-		// Go through each IPAM allocation, check its attributes for the node it is assigned to.
-		for _, idx := range b.Allocations {
-			if idx == nil {
-				// Not allocated.
-				continue
+	for poolName, poolBlocks := range c.blocksByPool {
+		// These counts track pool-based gauges by node for the current pool.
+		inUseAllocationsForPoolByNode := map[string]int{}
+		borrowedAllocationsForPoolByNode := map[string]int{}
+		blocksForPoolByNode := map[string]int{}
+		gcCandidatesForPoolByNode := map[string]int{}
+
+		for blockCIDR, _ := range poolBlocks {
+			b := c.allBlocks[blockCIDR].Value.(*model.AllocationBlock)
+
+			affineNode := "no_affinity"
+			if b.Affinity != nil {
+				affineNode = strings.TrimPrefix(*b.Affinity, "host:")
 			}
-			attr := b.Attributes[*idx]
 
-			// Track nodes based on IP allocations.
-			if node, ok := attr.AttrSecondary[ipam.AttributeNode]; ok {
+			legacyBlocksByNode[affineNode]++
+			blocksForPoolByNode[affineNode]++
+
+			// Go through each IPAM allocation, check its attributes for the node it is assigned to.
+			for _, allocation := range c.allocationsByBlock[blockCIDR] {
+				// Track nodes based on IP allocations.
+				allocationNode := allocation.node()
+				if allocationNode == "" {
+					allocationNode = "unknown_node"
+				}
+
 				// Update metrics maps with this allocation.
-				if b.Affinity == nil || node != strings.TrimPrefix(*b.Affinity, "host:") {
-					// If the allocations node doesn't match the block's, then this is borrowed.
-					borrowedIPsByNode[node]++
+				inUseAllocationsForPoolByNode[allocationNode]++
+
+				if allocationNode != "unknown_node" && (b.Affinity == nil || allocationNode != affineNode) {
+					// If the allocation's node doesn't match the block's, then this is borrowed.
+					legacyBorrowedIPsByNode[allocationNode]++
+					borrowedAllocationsForPoolByNode[allocationNode]++
+				}
+
+				// Update candidate count. Include confirmed leaks as well, in case there is an issue keeping them
+				// from being immediately reclaimed as usual.
+				if allocation.isCandidateLeak() || allocation.isConfirmedLeak() {
+					gcCandidatesForPoolByNode[allocationNode]++
 				}
 			}
 		}
+
+		// Update gauge values, retrieving and resetting the values for the current pool
+		inUseAllocationsGaugeForPool := inUseAllocationGauges[poolName]
+		inUseAllocationsGaugeForPool.Reset()
+		for node, uses := range inUseAllocationsForPoolByNode {
```

Comment:
hm, are you sure they need that form of expression? 

Does `sum(ipam_allocations_in_use{ippool="test"})` not return `0` if there are no values?

Reply:
My understanding is that the expression returns no data rather than 0.

Here is an [example](http://demo.robustperception.io:9090/graph?g0.expr=sum(prometheus_sd_failed_configs%7Bname%3D%22scrape%22%7D)&g0.tab=1&g0.stacked=0&g0.range_input=1h) of gauge returning zero, and [another](http://demo.robustperception.io:9090/graph?g0.expr=sum(prometheus_sd_failed_configs%7Bname%3D%22null%22%7D)&g0.tab=1&g0.stacked=0&g0.range_input=1h) with the same gauge using a non existent label

Reply:
seems like a reasonable idea to explicitly set this to zero then (and delete it altogether when the pool no longer exists, of course)

Reply:
I've implemented explicitly zeroed gauge metrics with my pushed changes. Two points I wanted to raise for discussion:

1. In our brief discussion on implementation in Slack, you had implied that we only zero the gauge for the "in use" metric. I decided to zero all 3 allocation gauges ("in use", "borrowed", "GC candidate") for consistency and to allow similar percentage expressions to be constructed by users on these metrics. I'm not sure if it was a conscious implication on your part that we should only zero the "in use" gauge - if so let me know and I can easily adjust.

2. We discussed using the existing mapping of calico nodes to kubernetes nodes in order to enumerate calico nodes for gauge zeroing. I discovered that this mapping does not add an entry if the node is orchestrated by kubernetes. Do you think we should be zeroing for non-kubernetes calico nodes as well?